### PR TITLE
Adds linking tickets in asay, referencing mob datums in asay, improves asay ref code

### PR DESCRIPTION
**Pull request by Ryll-Ryll**
## About The Pull Request
Continuing on the work started in #65154 and 61712, this PR expands what asay linking can be used for and improves the code for it. You can now link tickets in asay by saying `#[ticket ID]` like `#1`. This shows the ticket's status and who filed the ahelp. Might remove the ticket filer if it's not really useful.

This also fixes an oversight of mine with datum linking, with how mobs and certain other datums (armor datums, landmarks) don't follow the memory address format for their `tag` values. Datum linking now uses a `@[tag]` format, so you can link to a mob by saying `@mob_702` and such.

[](https://i.imgur.com/s5sEqQ8.png)

## Why It's Good For The Game
More robust admin tools

## Changelog

:cl: Ryll/Shaps
admin: You can now link tickets in asay with #[ticket ID], like #1 and #5.
admin: asay datum linking now functions by saying '@[datum's tag value]`, whether it's a normal datum (@0x2001169) or a mob (@mob_702) or whatever else.

---
---
---
---

[](https://i.imgur.com/Q9qoMdd.png)
(old pic)

### CODE REVIEW

**Comment by Fikou** [comment]
linking tickets is too easy to accidentally do

**Comment by Fikou** [comment]
> > linking tickets is too easy to accidentally do
> 
> why is this a problem

when im talking about number three i dont want an entire ticket to pop up

**Comment by magatsuchi** [comment]
> > > linking tickets is too easy to accidentally do
> > 
> > 
> > why is this a problem
> 
> when im talking about number three i dont want an entire ticket to pop up

[#n] would probably be a better sub then for linking

**Comment by Ryll-Ryll** [comment]
I can remove the ahelper's ckey+name from the ticket link and reduce "Active/Resolved/Closed" to "(A/R/C)" to make it less intrusive but adding brackets in addition to the number sign would suck

File: code/modules/admin/verbs/adminhelp.dm [hunk 1]
```diff
@@ -938,65 +938,74 @@ GLOBAL_DATUM_INIT(admin_help_ui_handler, /datum/admin_help_ui_handler, new)
...
-		if(!istype(check_datum))
-			continue
-		msglist[i] = "<u><a href='?_src_=vars;[HrefToken(TRUE)];Vars=[word_with_brackets]'>[word_with_brackets]</A></u>"
-		datums_to_ref[word] = word
+		switch(word[1])
+			if("@")
+				var/stripped_word = lowertext(copytext(word, 2))
+
+				// first we check if it's a ckey of an admin
+				var/client/client_check = GLOB.directory[stripped_word]
```
Comment (Watermelon914) [inline code/modules/admin/verbs/adminhelp.dm:969]: before passing it in here, you could also do ckey(stripped_word) to turn it into a ckey, which'll also get rid of any special characters that you can't have in a ckey but you might be able to have in a key

File: code/modules/admin/verbs/adminhelp.dm [hunk 2]
```diff
@@ -938,65 +938,74 @@ GLOBAL_DATUM_INIT(admin_help_ui_handler, /datum/admin_help_ui_handler, new)
...
+				if(client_check?.holder)
+					msglist[i] = "<u>[word]</u>"
+					pinged_admins[stripped_word] = client_check
+					modified = TRUE
+					continue
+
+				// then if not, we check if it's a datum ref
+
+				var/word_with_brackets = "\[[stripped_word]\]" // the actual memory address lookups need the bracket wraps
```
Comment (Watermelon914) [inline code/modules/admin/verbs/adminhelp.dm:978]: don't need the \] at the end here

**Comment by ShizCalev** [comment]
conflict

File: code/modules/admin/verbs/adminhelp.dm (hunk 1 above)
Comment (Ryll-Ryll) [reply code/modules/admin/verbs/adminhelp.dm:969]: good idea

File: code/modules/admin/verbs/adminhelp.dm (hunk 2 above)
Comment (Ryll-Ryll) [reply code/modules/admin/verbs/adminhelp.dm:978]: It's intentional, so that it shows "[tag]", matches the opening escaped bracket

**Pull request merged**